backfiller/backfiller/main.py

```diff
@@ -0,0 +1,227 @@
+"""Download segments from other nodes to catch stuff this node missed."""
+# TODO logging, better exception handling
+
+import datetime
+import errno
+import os
+import random
+import time
+import uuid
+
+import requests
+
+import common
+
+
+HOUR_FMT = '%Y-%m-%dT%H'
+TIMEOUT = 5 #default timeout for remote requests 
+
+
+def get_nodes():
+	"""List address of other wubloaders.
+	
+	This returns a list of the other wubloaders as strings of the form 'protocol://host:port/'"""
+	# either read a config file or query the database to get the addresses
+	# of the other nodes
+	# figure out some way that the local machine isn't in the list of returned
+	# nodes so that 
+	# as a prototype can just hardcode some addresses.
+
+	nodes = []
+	return nodes
+
+
+def list_local_segments(base_dir, stream, variant, hour):
+	"""List segments in a given hour directory.
+
+	For a given base_dir/stream/variant/hour directory return a list of non-hidden files. If the directory path is not found, return an empty list. 
+
+	Based on based on restreamer.list_segments. We could just call restreamer.list_segments but this avoids HTTP/JSON overheads."""
+
+	path = os.path.join(base_dir, stream, variant, hour)
+	try:
+		return [name for name in os.listdir(path) if not name.startswith('.')]
+
+	except OSError as e:
+		if e.errno != errno.ENOENT:
+			raise
+		return []	
+
+
+def list_remote_hours(node, stream, variant, timeout=TIMEOUT):
+	"""Wrapper around a call to restreamer.list_hours."""
+	uri = '{}/files/{}/{}'.format(node, stream, variant)
+	resp = requests.get(uri, timeout=timeout)
+	return resp.json()
+
+
+def list_remote_segments(node, stream, variant, hour, timeout=TIMEOUT):
+	"""Wrapper around a call to restreamer.list_segments."""
+	uri = '{}/files/{}/{}/{}'.format(node, stream, variant, hour)
+	resp = requests.get(uri, timeout=timeout)
+	return resp.json()
+
+
+def get_remote_segment(base_dir, node, stream, variant, hour, missing_segment,
+			timeout=TIMEOUT):
+	"""Get a segment from a node.
+
+	Fetches stream/variant/hour/missing_segment from node and puts it in base_dir/stream/variant/hour/missing_segment. If the segment already exists locally, this does not attempt to fetch it."""
+
+	path = os.path.join(base_dir, stream, variant, hour, missing_segment)
+	# check to see if file already exists to avoid unnecessarily copying it
```

Comment:
expand comment a little: we already checked once when we got a list of "missing segments", why are we checking again? eg. `file might have been created since we listed our segments, check again to see if we don't need to copy it after all`